FieldOpt/Optimization/optimizers/CMA_ES.cpp

```diff
@@ -0,0 +1,334 @@
+}
+
+vector<CMA_ES::Individual> CMA_ES::sortPopulation(vector<CMA_ES::Individual> population) {
+    vector<CMA_ES::Individual> sortedPopulation;
+    int n = population_.size();
+    bool swapped = true;
+    while (swapped) {
+        swapped = false;
+        for (int i = 1; i <= n - 1; i++) {
+            if (population[i - 1].ofv() > population[i].ofv() &&
+                (Settings::Optimizer::Minimize == settings_->mode())) {
+                CMA_ES::Individual temp_individual = population[i];
+                population[i] = population[i - 1];
+                population[i - 1] = temp_individual;
+                swapped = true;
+            } else if (population[i - 1].ofv() < population[i].ofv() &&
+                       (Settings::Optimizer::Maximize == settings_->mode())) {
+                CMA_ES::Individual temp_individual = population[i];
+                population[i] = population[i - 1];
+                population[i - 1] = temp_individual;
+                swapped = true;
+            }
+        }
+    }
+    return population;
+}
+
+void CMA_ES::iterate() {
+    if (enable_logging_) {
+        logger_->AddEntry(this);
+    }
+    for (int i = 0; i < population_.size(); i++) {
+        if (Settings::Optimizer::Minimize == settings_->mode()) {
+            population_[i].case_pointer_->set_objective_function_value(
+                    population_[i].ofv() + (exp(penalty_ * population_[i].penalty_dist_) - 1));
+        } else {
+            population_[i].case_pointer_->set_objective_function_value(
+                    population_[i].ofv() - (exp(penalty_ * population_[i].penalty_dist_) - 1));
+        }
+    }
+    population_ = sortPopulation(population_);
+    xold_ = xmean_;
+    xmean_ = Eigen::VectorXd::Zero(xmean_.size());
+    for (int i = 0; i < n_vars_; i++) {
+        double temp = 0;
+        for (int j = 0; j < mu_; j++) {
+            temp += population_[j].erands_norm_[i] * weights_[j];
+        }
+        xmean_(i) = temp;
+    }
+    updateEvolutionPath();
+    adaptCovarianceMatrix();
+    decompositionOfC();
+
+    for (int i = 0; i < lambda_; ++i) {
+        auto new_case = generateCase(xmean_, i, 0);
+        case_handler_->AddNewCase(new_case);
+    }
+    population_ = temp_population_;
+    iteration_++;
+}
+
+void CMA_ES::handleEvaluatedCase(Case *c) {
+    if (isImprovement(c)) {
+        updateTentativeBestCase(c);
+        if (VERB_OPT > 0) {
+            stringstream ss;
+            ss.precision(6);
+            ss << scientific;
+            ss << "New best in population, iteration " << Printer::num2str(iteration_) << ": OFV "
+               << c->objective_function_value();
+            Printer::ext_info(ss.str(), "Optimization", "CMA_ES");
+        }
+    }
+}
+
+Optimizer::TerminationCondition CMA_ES::IsFinished() {
+    if (case_handler_->CasesBeingEvaluated().size() > 0) return NOT_FINISHED;
+    if (iteration_ < max_iterations_) return NOT_FINISHED;
+    else return MAX_EVALS_REACHED;
```

Comment:
You should probably add a `logger_->AddEntry(this);`, otherwise you run the risk of the best case not being logged if it is found in the final iteration.